File: src/MagnusAddon/language_services/janome_ex/tokenizing/pre_processing_stage/dictionary_form_verb_splitter.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots
from language_services.janome_ex.tokenizing.godan_dictionary_form_stem import GodanDictionaryFormInflection, GodanDictionaryFormStem, IchidanDictionaryFormInflection, IchidanDictionaryFormStem, KuruVerbDictionaryFormInflection, KuruVerbDictionaryFormStem, SuruVerbDictionaryFormInflection, SuruVerbDictionaryFormStem

if TYPE_CHECKING:
    from language_services.janome_ex.tokenizing.analysis_token import IAnalysisToken
    from language_services.janome_ex.tokenizing.jn_token import JNToken

class DictionaryFormVerbSplitter(Slots):
    @classmethod
    def try_split(cls, token: JNToken) -> list[IAnalysisToken] | None:
        if token.is_dictionary_form() and not token.is_progressive_form():
            if token.is_ichidan_verb:
                return cls.split_ichidan_dictionary_form(token)

            if token.is_godan_verb:
                return cls.split_godan_dictionary_form(token)

            if token.is_kuru_verb:
                return cls.split_kuru_verb(token)

            if token.is_suru_verb:
                return cls.split_suru_verb(token)

        return None

    @classmethod
    def split_ichidan_dictionary_form(cls, token: JNToken) -> list[IAnalysisToken]:
        ichidan_surface = token.surface[:-1]
        return [IchidanDictionaryFormStem(token, surface=ichidan_surface, base=token.base_form),
                IchidanDictionaryFormInflection(token)]

    @classmethod
    def split_godan_dictionary_form(cls, token: JNToken) -> list[IAnalysisToken]:
        godan_surface = token.surface[:-1]
        godan_dictionary_ending = token.surface[-1]
        return [GodanDictionaryFormStem(token, surface=godan_surface, base=token.base_form),
                GodanDictionaryFormInflection(token, surface=godan_dictionary_ending, base="う")]

    @classmethod
    def split_kuru_verb(cls, token: JNToken) -> list[IAnalysisToken]:
        surface = token.surface[:-1]
        return [KuruVerbDictionaryFormStem(token, surface=surface, base=token.base_form),
                KuruVerbDictionaryFormInflection(token)]

    @classmethod
    def split_suru_verb(cls, token: JNToken) -> list[IAnalysisToken]:
        surface = token.surface[:-1]
        return [SuruVerbDictionaryFormStem(token, surface=surface, base=token.base_form),
                SuruVerbDictionaryFormInflection(token)]

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/PreProcessingStage/DictionaryFormVerbSplitter.cs
using System.Collections.Generic;

namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing.PreProcessingStage;

public static class DictionaryFormVerbSplitter
{
    public static List<IAnalysisToken>? TrySplit(JNToken token)
    {
        if (token.IsDictionaryForm() && !token.IsProgressiveForm())
        {
            if (token.IsIchidanVerb)
            {
                return SplitIchidanDictionaryForm(token);
            }

            if (token.IsGodanVerb)
            {
                return SplitGodanDictionaryForm(token);
            }

            if (token.IsKuruVerb)
            {
                return SplitKuruVerb(token);
            }

            if (token.IsSuruVerb)
            {
                return SplitSuruVerb(token);
            }
        }

        return null;
    }

    static List<IAnalysisToken> SplitIchidanDictionaryForm(JNToken token)
    {
        var ichidanSurface = token.Surface[..^1];
        return
        [
           new IchidanDictionaryFormStem(token, ichidanSurface, token.BaseForm),
           new IchidanDictionaryFormInflection(token)
        ];
    }

    static List<IAnalysisToken> SplitGodanDictionaryForm(JNToken token)
    {
        var godanSurface = token.Surface[..^1];
        var godanDictionaryEnding = token.Surface[^1..];
        return
        [
           new GodanDictionaryFormStem(token, godanSurface, token.BaseForm),
           new GodanDictionaryFormInflection(token, godanDictionaryEnding, "う")
        ];
    }

    static List<IAnalysisToken> SplitKuruVerb(JNToken token)
    {
        var surface = token.Surface[..^1];
        return
        [
           new KuruVerbDictionaryFormStem(token, surface, token.BaseForm),
           new KuruVerbDictionaryFormInflection(token)
        ];
    }

    static List<IAnalysisToken> SplitSuruVerb(JNToken token)
    {
        var surface = token.Surface[..^1];
        return
        [
           new SuruVerbDictionaryFormStem(token, surface, token.BaseForm),
           new SuruVerbDictionaryFormInflection(token)
        ];
    }
}
